boltdb/boltdb_manager.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package boltdb
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/ts-bridge/storage"
+	"os"
+	"path"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/timshannon/bolthold"
+)
+
+// Options holds settings specific to boltdb
+type Options struct {
+	DBPath string
+}
+
+// Manager struct implementing the storage.Manager interface
+type Manager struct {
+	Store *bolthold.Store
+}
+
+// New initializes the Manager struct implementing a generic storage.Manager interface
+func New(_ context.Context, options *Options) *Manager {
+	// If path for a DB is not configured, default to working directory.
+	boltPath := options.DBPath
+	if boltPath == "" {
+		pwd, err := os.Getwd()
+		if err != nil {
+			log.Fatalf("couldn't get app working directory %v", err)
+		}
+		boltPath = path.Join(pwd + "/bolt.db")
+	}
+	store, err := bolthold.Open(boltPath, 0666, nil)
+	if err != nil {
+		log.Fatalf("Unable to Open BoltDB at %v:%v", boltPath, err)
+	}
+	log.Debug("Opened BoltDB")
+
+	return &Manager{Store: store}
+}
+
+// NewMetricRecord returns a BoltDB-based metric record for a given metric name
+func (d *Manager) NewMetricRecord(_ context.Context, name, query string) (storage.MetricRecord, error) {
+	r := &StoredMetricRecord{Name: name, storage: d}
+	if err := r.load(); err != nil {
+		return nil, err
+	}
+	r.Query = query
+	return r, nil
+}
+
+// CleanupRecords removes obsolete metric records from BoltDB.
+//   `keep` represents metrics to be kept, all others will be purged
+func (d *Manager) CleanupRecords(ctx context.Context, keep []string) error {
+	// datatype is just an example of the type stored so that the proper bucket and indexes are updated
+	datatype := StoredMetricRecord{}
+	query := bolthold.Where("Name").Not().In(bolthold.Slice(keep)...)
```

Comment:
Perhaps this should filter by key (`bolthold.Key`) rather than the `Name` field? I know they are the same in practice, but if we decide to change the keys in the future this could get out of sync.

Reply:
That's a very good point, didn't think of that. Changed, thanks! :)